src/main/java/com/epam/aidial/core/upstream/UpstreamRouteProvider.java

```diff
@@ -0,0 +1,111 @@
+package com.epam.aidial.core.upstream;
+
+import com.epam.aidial.core.config.Addon;
+import com.epam.aidial.core.config.Application;
+import com.epam.aidial.core.config.Assistant;
+import com.epam.aidial.core.config.Config;
+import com.epam.aidial.core.config.Deployment;
+import com.epam.aidial.core.config.Model;
+import com.epam.aidial.core.config.Route;
+import com.epam.aidial.core.config.Upstream;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@Slf4j
+public class UpstreamRouteProvider {
+
+    /**
+     * Cached load balancers for config deployments
+     */
+    private volatile Map<String, TieredBalancer> balancers = new HashMap<>();
+
+    /**
+     * Returns UpstreamRoute for the given provider
+     *
+     * @param provider upstream provider for any deployment with actual upstreams
+     * @return upstream route
+     */
+    public UpstreamRoute get(UpstreamProvider provider) {
+        String deploymentName = provider.getName();
+        List<Upstream> upstreams = provider.getUpstreams();
+
+        TieredBalancer balancer = balancers.get(deploymentName);
+        if (balancer == null) {
+            // if no state found for upstream, it's probably custom application
+            balancer = new TieredBalancer(deploymentName, upstreams);
+        }
+
+        return new UpstreamRoute(balancer, 5);
```

Comment:
Where does the number 5 come from? Shouldn't we use the number of upstreams instead?

Reply:
no, this is basically max retry attempts, we don't want to stick this number to upstream count, because applications usually have only 1 upstream

Reply:
Please put the value to constant with description